File: review/subsets-ii.ts
/**
 * https://leetcode.com/problems/subsets-ii/description/
 */
export { subsetsWithDup };

function subsetsWithDup(nums: number[]): number[][] {
  nums.sort((a, b) => a - b);
  const result: number[][] = [];

  function backtrack(index: number, path: number[]) {
    result.push([...path]);

    for (let i = index; i < nums.length; i++) {
      if (i > index && nums[i] === nums[i - 1]) continue;
      path.push(nums[i]);
      backtrack(i + 1, path);
      path.pop();
    }
  }

  backtrack(0, []);
  return result;
}

File: returned/python/today6/subsets_ii.py
# https://leetcode.com/problems/subsets-ii/description/
from typing import List

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        
        def backtrack(i, subset):
            if i == len(nums):
                result.append(subset[::])
                return
            
            subset.append(nums[i])
            backtrack(i+1, subset)
            subset.pop()

            while i + 1 < len(nums) and nums[i] == nums[i+1]:
                i += 1
            backtrack(i+1, subset)
        
        backtrack(0, [])
        return result
